### PR TITLE
Run LocalPackageSearchResource search in Long running thread

### LINKED ISSUE #11570: [Responsiveness] Package Management UI can consume large number of threads all searching the disk, it needs to run from long running thread
Tracking issue for: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1424937
I added `Around The World` tag here since internal issue has one.
Currently [LocalPackageSearchResource.SearchAsync](https://github.com/NuGet/NuGet.Client/blob/ac2cb86d726e1f0089fc6f5a63357a37940ca8aa/src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalPackageSearchResource.cs#L30) it's blocking (up to 316 threads from Thread.Pool which is designed for short/fast operations, instead we need to run on dedicated long running thread so we don't block Thread.Pool.

Sample look:
LockingThreads

### PR DESCRIPTION
**Pull request by erdembayar**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11570

Regression? Last working version:

## Description
In general Task.Run borrow thread from pre-populated Thread.Pool and it's intended for very fast operation then return. But some local V2 could be very slow due to its being UNC path over network or very large/deep layered folders. So instead of running on thread from Thread.Pool we can run it using dedicated long running thread.
I'll create another follow up PR for passing cancellation token to thread then check frequently and iterate directories one by one instead of trying to get all once.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Comment by erdembayar** [comment]
FYI: @davkean 
It's initial pass for making long running thread.

File: src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalPackageSearchResource.cs [hunk 1]
```diff
@@ -35,7 +35,7 @@ public LocalPackageSearchResource(FindLocalPackagesResource localResource)
             ILogger log,
             CancellationToken token)
         {
-            return await Task.Run(() =>
+            return await Task.Factory.StartNew(() =>
```
Comment (davkean) [inline src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalPackageSearchResource.cs:38]: In the future, you can just remove these "await" and just return Task.StartNew directly to remove unneeded async state machines. While here it probably doesn't matter, but elsewhere in NuGet they do show up as overhead.

**Review by davkean — APPROVED** [review]

**Comment by davkean** [comment]
For context for reviewers, you can see these this in the failure in #1424937 internally. You want tasks that run on the thread pool to be very short and non-blocking, this path synchronously hits the disk and under utilizes the CPU, hence the TaskCreationOptions.LongRunning which creates a dedicated thread.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs [hunk 2]
```diff
@@ -586,18 +587,53 @@ public async Task LocalPackageSearchResource_FileSourceAsync()
...
+                // Act & Assert
+                await Assert.ThrowsAsync<TaskCanceledException>(
                     async () => await resource.SearchAsync("", null, 0, 1, NullLogger.Instance, new CancellationToken(canceled: true)));
             }
         }
 
+        [Fact]
+        public async Task LocalPackageSearch_SearchAsync_SlowLocalRepository_WithCancellationToken_ThrowsAsync()
+        {
+            using (var pathContext = new SimpleTestPathContext())
```
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs]: Can you use the inline `using` here as well?

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs [hunk 3]
```diff
@@ -586,18 +587,53 @@ public async Task LocalPackageSearchResource_FileSourceAsync()
...
+                       {
+                           Thread.Sleep(5000);
+                           cancellationToken.ThrowIfCancellationRequested();
+                       }
+                       ).Returns(Enumerable.Empty<LocalPackageInfo>);
+                LocalPackageSearchResource resource = new LocalPackageSearchResource(localRepository.Object);
+
+                // Act
+                Task delay = Task.Delay(TimeSpan.FromMilliseconds(500), cts.Token);
+                Task searchTask = resource.SearchAsync("", null, 0, 1, NullLogger.Instance, cts.Token);
```
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs]: Please name parameters

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs [hunk 4]
```diff
@@ -586,18 +587,53 @@ public async Task LocalPackageSearchResource_FileSourceAsync()
...
+
+                // Assert
+                var completed = await Task.WhenAny(searchTask, delay);
```
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs]: use explicit type

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs [hunk 5]
```diff
@@ -586,18 +587,53 @@ public async Task LocalPackageSearchResource_FileSourceAsync()
...
+                if (completed != delay)
```
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs]: I see this is because 5000 > 500 milliseconds. Can you add a comment? It's not obvious why "completed != delay" warrants a timeout. 
Alternatively, a message in the exception to explain the situation would be good. Otherwise, if I saw this in a test failure, I'd think we had some bug with timeout logic.

**Review by donnie-msft — COMMENTED** [review]
Cool! here's some nits

File: src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalPackageSearchResource.cs (hunk 1 above)
Comment (zivkan) [reply src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalPackageSearchResource.cs:38]: Note that this directly contradicts David Fowler's advice: https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task

He does note that there are perf considerations. The advantage I personally find most compelling is in methods that do some work before calling (and returning) an async method's Task, it will change exception behaviour. `async` methods will always return an `AggregateException`, whereas a non-async method will throw the "real" exception directly, which might break the logic of callers excepection handling expectations (always get an AggregateException).

In this particular case where there's no additional logic, therefore no chance of exception, just calling another Async method and returning immediately, there's little downside to removing the async keyword from the method signature. But I would need convincing to treat "change all `return await` to `return` without `async` on the method signature" to be a rule of thumb to always follow.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs (hunk 2 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs]: Sure.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs (hunk 5 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs]: Sure, I added comments. Please check if it's clear for someone reading code later.

File: src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalPackageSearchResource.cs (hunk 1 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalPackageSearchResource.cs:38]: Sure, I'll try to keep number of machine states down for new code.
But I'll keep as it's now, there're quite many overrides share signature with this method, we can revisit it later if needed.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalPackageSearchResource.cs:38]: Removing async/await doesn't change the signature. 

I'm not saying do, just that what you're being asked to do isn't a signature change.

Comment (erdembayar) [reply src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalPackageSearchResource.cs:38]: Ok. I'll keep it for now, I'll think more about it in next iteration.

**Review by nkolev92 — APPROVED** [review]
The src code looks good. 

I'm not sure the test is helpful right now.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs [hunk 6]
```diff
@@ -586,18 +587,57 @@ public async Task LocalPackageSearchResource_FileSourceAsync()
...
 
-                await Assert.ThrowsAsync<OperationCanceledException>(
+                // Act & Assert
+                await Assert.ThrowsAsync<TaskCanceledException>(
                     async () => await resource.SearchAsync("", null, 0, 1, NullLogger.Instance, new CancellationToken(canceled: true)));
             }
         }
 
+        [Fact]
+        public async Task LocalPackageSearch_SearchAsync_SlowLocalRepository_WithCancellationToken_ThrowsAsync()
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs]: I meant to leave a comment here.

I don't think this test is helpful right now. 

The cancellation check is raised by test code, not the product code, so you're effective testing the test code.

Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs]: It's best I can do for now, idea is same. I'll try to come up with something better for next PR.

Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs]: Let me rephrase. 
I don't think it adds anything, and I think it should be removed.

You can add an equivalent test in the future PR you mentioned.

Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs]: Sure, removed.
